pkg/tunneler/integration_test.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tunneler
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/http/httputil"
+	"net/url"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// withClusterScope mock the cluster name
+func withClusterScope(apiHandler http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		var cluster request.Cluster
+		cluster.Name = logicalcluster.New("test")
+		ctx := request.WithCluster(req.Context(), cluster)
+		apiHandler.ServeHTTP(w, req.WithContext(ctx))
+	}
+}
+
+func setup(t *testing.T) (*http.Client, string, func()) {
+	t.Helper()
+	backend := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Hello world")
+	}))
+	backend.EnableHTTP2 = true
+	backend.StartTLS()
+
+	// public server
+	mux := http.NewServeMux()
+	apiHandler := WithSyncerTunnel(mux)
+	apiHandler = withClusterScope(apiHandler)
+	publicServer := httptest.NewUnstartedServer(apiHandler)
+	publicServer.EnableHTTP2 = true
+	publicServer.StartTLS()
+
+	// private server
+	l, err := NewListener(publicServer.Client(), publicServer.URL, "d001")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// reverse proxy queries to an internal host
+	url, err := url.Parse(backend.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	proxy.Transport = backend.Client().Transport
+	server := &http.Server{Handler: proxy}
+	//nolint:errcheck
+	go server.Serve(l)
+
+	// client
+	// wait for the reverse connection to be established
+	time.Sleep(1 * time.Second)
```

Comment:
why a second? Can this flake?

Reply:
I run all my tests with `go race -c` and then `stress` to check they don't flake , I can't remember now if I did it with last versions, but I'll do it befor emerge

Reply:
https://github.com/kcp-dev/kcp/pull/873/commits/ca18e5b3e6dd09216c294fd93cdc2e4e4350e049#r960457232

Reply:
what I mean is: if you didn't wait, would it flake?

Reply:
yes, the listener needs to connect to the handler so the reverse connection is established before start testing

We can do an active check instead of the time.sleep , but this is reliable and scoped to this setup() function only ... I'll add it to my TODO backlog